repository/src/main/java/org/alfresco/repo/virtual/template/PagingRequestConstraint.java

```diff
@@ -55,7 +55,17 @@ protected SearchParameters applyDecorations(ActualEnvironment environment, Searc
         if (pagingRequest != null)
         {
             searchParametersCopy.setSkipCount(pagingRequest.getSkipCount());
-            searchParametersCopy.setMaxItems(pagingRequest.getMaxItems());
+
+            // MNT-20791 : Smart folders will retrieve incorrect totalResults.
+            // Search query will return a set of size equal to searchParameters.maxItems
+            // For pagingRequest maxItems represents number of elements on the page and requestTotalCountMax the maximum number of elements to retrieve
+            // The Search query will retrieve elements starting from skipCount + 1, because of this in order to display pages consisten in share
+            // the numberOfElementsToRetrieve will be requestTotalCountMax - skipCount.
+            int numberOfElementsToRetrieve = pagingRequest.getRequestTotalCountMax() != 0
+                ? pagingRequest.getRequestTotalCountMax() - pagingRequest.getSkipCount()
```

Comment:
I don't fully understand this change.
The whole point of paging is to query only fractions of the entire set for each page. If I have 200 items in the result set, then I am only retrieving items 1..50 for the first page. Then on the second page items 51..100, and so on.
With this change, I am loading items 1..200 for the first page, items 51..200 for the second page, etc.
Then why having paging at all? With this change in place, it would be more efficient to not use paging at all and just load all items once.